src/clj/bluegenes_tool_store/tools.clj

```diff
@@ -46,33 +54,93 @@
              :npm (get-in package [:name])}
      :config config
-     :package package
+     :package (select-keys package [:description :license :homepage :name :author :version])
       ;; return image path if it exists, or false otherwise.
-     :hasimage (if (.exists (io/file (str path preview-image)))
+     :hasimage (if (.exists (io/file path preview-image))
                  browser-path
                  false)}))
 
+;; A valid tool needs to have the following:
+;; - package.json
+;; - config.json
+;; - dist/bundle.js
+
 (defn installed-tools-list
   "Return a list of the installed tools listed in the package.json file. "
   []
-  (try
-    (log :info "|-- looking for tool config file at: " tools-config)
-    (let [packages (cheshire/parse-string (slurp tools-config) true)
-          package-names (keys (:dependencies packages))]
-      package-names)
-    (catch Exception e (log :warn
-                            (str "Couldn't find tools at " tools-config (.getMessage e) "- please run `npm init -y` in the tools directory and install your tools again.")))))
+  (let [config-path (tools-path :config)]
+    (try
+     (let [packages (cheshire/parse-string (slurp config-path) true)
+           package-names (keys (:dependencies packages))]
+       package-names)
+     (catch Exception e
+            (log :warn
+                 (str "Couldn't find tools at " config-path (.getMessage e) "- please run `npm init -y` in the tools directory and install your tools again."))))))
+
+(defn tools-list-res
+  "Create response containing the list of tools."
+  []
+  (let [modules-path (tools-path :modules)]
+    (response/ok
+      {:tools (map #(tool-config % modules-path)
+                   (installed-tools-list))})))
+
+(defn get-all-tools
+  "Return list of tools as a REST response to our GET."
+  []
+  (tools-list-res))
 
-(defn tools
-  "Format the list of tools as a REST response to our GET."
+(defn get-tools-path
+  "Respond with tool path."
   []
-  (log :info "Tools folder: ")
-  (log :info "|--" tool-path)
-  (let [installed-tools (installed-tools-list)
-        res
-        {:tools
-         (reduce
-          (fn [tool-list newitem]
-            (conj tool-list (get-tool-config newitem tool-path)))
-          #{} installed-tools)}]
-    (response/ok res)))
+  (response/ok {:path (:bluegenes-tool-path env)}))
+
+(let [lock (ReentrantLock.)]
+  (defn sync-sh-req
+    "Run a synchronous shell command, rejecting in the case where a command
+    is already in progress. This is mostly for wrapping calls to npm, which
+    only allows a single process at a time.
+    Returns a response with the new tool list on success."
+    [cmdv dir]
+    (if (.tryLock lock)
+      (try
+        (with-sh-dir dir
+          (apply sh cmdv))
+        (tools-list-res)
+        (finally
+          (.unlock lock)))
+      (response/service-unavailable {:error "A tool operation is already in progress. Please try again later."}))))
+
+(defn install-package
```

Comment:
I suspect we might need more error handling around the install/update functions in this file, too.